src/frontends/lean/inductive_cmds.cpp

```diff
@@ -360,36 +360,52 @@ class inductive_cmd_fn {
         }
     }
 
-    void parse_intro_rules(bool has_params, expr const & ind, buffer<expr> & intro_rules, bool prepend_ns) {
-        // If the next token is not `|`, then the inductive type has no constructors
-        if (m_p.curr_is_token(get_bar_tk())) {
-            m_p.next();
-            while (true) {
-                m_pos = m_p.pos();
-                name ir_name = mlocal_name(ind) + m_p.check_atomic_id_next("invalid introduction rule, atomic identifier expected");
-                if (prepend_ns)
-                    ir_name = get_namespace(m_env) + ir_name;
-                parser::local_scope S(m_p);
-                buffer<expr> params;
-                implicit_infer_kind kind = implicit_infer_kind::Implicit;
-                m_p.parse_optional_binders(params, kind);
-                m_implicit_infer_map.insert(ir_name, kind);
-                for (expr const & param : params)
-                    m_p.add_local(param);
-                expr ir_type;
-                if (has_params || m_p.curr_is_token(get_colon_tk())) {
-                    m_p.check_token_next(get_colon_tk(), "invalid introduction rule, ':' expected");
-                    ir_type = m_p.parse_expr();
+    void parse_intro_rules(bool has_params, expr const & ind, buffer<expr> & intro_rules,
+                           buffer<optional<std::string>> & intro_rule_docs, bool prepend_ns) {
+        // If the next token is neither `|` nor a doc block, then the inductive type has no constructors
+        if (!m_p.curr_is_token(get_bar_tk()) && (m_p.curr() != token_kind::DocBlock)) {
+            return;
+        }
+
+        while (true) {
+            optional<std::string> doc {};
+            if (m_p.curr() == token_kind::DocBlock) {
+                // If the next token is a doc block, it applies to the following constructor, *if one exists*.
+                // Otherwise, it might be part of the next command.
+                if (m_p.ahead_is_token(get_bar_tk()) || m_p.ahead_is_token(get_comma_tk())) {
+                    doc = m_p.parse_doc_block();
                 } else {
-                    ir_type = ind;
+                    return;
                 }
-                ir_type = Pi(params, ir_type, m_p);
-                intro_rules.push_back(mk_local(ir_name, ir_type));
-                lean_trace(name({"inductive", "parse"}), tout() << ir_name << " : " << ir_type << "\n";);
-                if (!m_p.curr_is_token(get_bar_tk()) && !m_p.curr_is_token(get_comma_tk()))
-                    break;
-                m_p.next();
+            } else if (!m_p.curr_is_token(get_bar_tk()) && !m_p.curr_is_token(get_comma_tk())) {
+                // No more constructors, we're done.
+                return;
```

Comment:
I'm not a fan of burying a `return` deep inside the control flow. Can we at least use a `break` instead?